File: scripts/python/src/tools/dotfile/packages.py
import os

from tools.core import blocks
from tools.core.dotfmt import formatted
from tools.dotfile.state import die, log, manifest_groups, sorted_entries

DEFAULT_GROUPS = [
    "shared",
    "linux/common",
    "linux/arch",
    "linux/ubuntu",
    "linux/kde",
    "linux/hyprland",
    "linux/server",
    "macos",
]

GROUP_CHARS = set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789._/-")
PACKAGE_CHARS = set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789._+@-")


def read_package_entries(ctx):
    try:
        return blocks.read(ctx.packages_config, comments=False, open_suffix=" {")
    except blocks.BlockError as error:
        die(blocks.describe(error, "config/packages.dotfile", "group"))
        return []


def load_package_metadata(ctx):
    ctx.package_descriptions = {}
    if not os.path.isfile(ctx.packages_config):
        return
    for entry in read_package_entries(ctx):
        if entry.opens:
            if not entry.block:
                die(f"config/packages.dotfile:{entry.number}: empty group")
            if set(entry.block) - GROUP_CHARS:
                die(f"config/packages.dotfile:{entry.number}: invalid group: {entry.block}")
            continue
        name, description = entry.split(" = ")
        if not name:
            die(f"config/packages.dotfile:{entry.number}: empty package")
        if set(name) - PACKAGE_CHARS:
            die(f"config/packages.dotfile:{entry.number}: invalid package: {name}")
        key = f"{entry.block}/{name}"
        if key in ctx.package_descriptions:
            die(f"config/packages.dotfile:{entry.number}: duplicate package: {key}")
        ctx.package_descriptions[key] = description


def package_groups(ctx):
    groups = list(DEFAULT_GROUPS)
    if os.path.isdir(ctx.environment_dir):
        manifests = []
        for parent, _dirnames, filenames in os.walk(ctx.environment_dir):
            if "manifest" in filenames:
                manifests.append(os.path.join(parent, "manifest"))
        for manifest in sorted(manifests):
            groups.extend(manifest_groups(manifest))
    seen = set()
    unique = []
    for group in groups:
        if group and group not in seen:
            seen.add(group)
            unique.append(group)
    return unique


def group_packages(directory):
    return [
        name for name in sorted_entries(directory) if os.path.isdir(os.path.join(directory, name))
    ]


def validate_package_names(ctx):
    for group in package_groups(ctx):
        directory = os.path.join(ctx.root, group)
        if not os.path.isdir(directory):
            continue
        for pkg in group_packages(directory):
            if pkg == "overrides":
                continue
            if set(pkg) - PACKAGE_CHARS:
                die(f"package directory has an unsupported name: {group}/{pkg}")


def render_packages(ctx):
    config_parts = []
    doc_parts = []
    wrote_group = False
    for group in package_groups(ctx):
        directory = os.path.join(ctx.root, group)
        if not os.path.isdir(directory):
            continue
        packages = [pkg for pkg in group_packages(directory) if pkg != "overrides"]
        if not packages:
            continue
        if wrote_group:
            config_parts.append("\n")
        config_parts.append(f"{group} {{\n")
        doc_parts.append(f"\n## `{group}`\n\n")
        for pkg in packages:
            description = ctx.package_descriptions.get(f"{group}/{pkg}", "")
            if description:
                config_parts.append(f"  {pkg} = {description}\n")
                doc_parts.append(f"- `{pkg}` — {description}\n")
            else:
                config_parts.append(f"  {pkg}\n")
                doc_parts.append(f"- `{pkg}`\n")
        config_parts.append("}\n")
        wrote_group = True
    # `dotfmt` owns the `=` column, so this only has to emit the entries.
    return formatted("".join(config_parts), ctx.packages_config), "".join(doc_parts)


def replace_package_file(content, destination, label):
    if os.path.isfile(destination):
        with open(destination, encoding="utf-8") as handle:
            if handle.read() == content:
                return 0
    with open(destination, "w", encoding="utf-8") as handle:
        handle.write(content)
    log(f"updated {label}")
    return 1


def sync_packages(ctx):
    validate_package_names(ctx)
    config, doc = render_packages(ctx)
    changed = replace_package_file(config, ctx.packages_config, "config/packages.dotfile")
    changed += replace_package_file(doc, ctx.packages_doc, "PACKAGES.md")
    return changed


def cmd_packages(ctx):
    load_package_metadata(ctx)
    if sync_packages(ctx) == 0:
        log("packages are current")

File: scripts/rust/crates/dotfile/src/artifacts/packages.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};

pub use crate::config::sorted_directories as directories;
use crate::config::{blocks, read_manifest};
use crate::context::{Context, write_atomic};
use crate::event::{Action, Event, EventSink, Phase};

pub const DEFAULT_GROUPS: &[&str] = &[
    "shared",
    "linux/common",
    "linux/arch",
    "linux/ubuntu",
    "linux/kde",
    "linux/hyprland",
    "linux/server",
    "macos",
];

pub fn synchronize(
    context: &Context,
    dry_run: bool,
    events: &dyn EventSink,
) -> Result<usize, String> {
    events.emit(Event::PhaseStarted {
        phase: Phase::Artifacts,
        total: Some(2),
    });
    let metadata = load_metadata(&context.packages_config)?;
    let groups = package_groups(context)?;
    validate_packages(context, &groups)?;
    let (config, document) = render(context, &groups, &metadata)?;
    let outputs = [
        (&context.packages_config, config, "package index"),
        (&context.packages_doc, document, "package documentation"),
    ];
    let mut changed = 0;
    for (index, (path, content, detail)) in outputs.into_iter().enumerate() {
        let differs = match fs::read(path) {
            Ok(current) => current != content.as_bytes(),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => true,
            Err(error) => return Err(format!("read {}: {error}", path.display())),
        };
        if differs {
            changed += 1;
            if !dry_run {
                write_atomic(path, content.as_bytes())?;
            }
        }
        events.emit(Event::Item {
            action: Action::Generate,
            path: path.clone(),
            detail: detail.to_string(),
            changed: differs,
        });
        events.emit(Event::Progress {
            phase: Phase::Artifacts,
            completed: index + 1,
            total: Some(2),
            label: detail.to_string(),
        });
    }
    Ok(changed)
}

pub fn load_metadata(path: &Path) -> Result<BTreeMap<String, String>, String> {
    let mut metadata = BTreeMap::new();
    for entry in blocks::read(path)? {
        let error = |message: String| format!("{}:{}: {message}", path.display(), entry.number);
        validate_group(&entry.block).map_err(&error)?;
        if entry.opens {
            continue;
        }
        let (name, description) = entry.split();
        validate_package(name).map_err(&error)?;
        let key = format!("{}/{name}", entry.block);
        if metadata
            .insert(key.clone(), description.to_owned())
            .is_some()
        {
            return Err(error(format!("duplicate package: {key}")));
        }
    }
    Ok(metadata)
}

pub fn package_groups(context: &Context) -> Result<Vec<String>, String> {
    let mut groups = DEFAULT_GROUPS
        .iter()
        .map(|group| (*group).to_string())
        .collect::<Vec<_>>();
    let mut manifests = Vec::new();
    collect_named_files(&context.environment_dir, "manifest", &mut manifests)?;
    manifests.sort();
    for manifest in manifests {
        groups.extend(read_manifest(&manifest)?);
    }
    let mut seen = BTreeSet::new();
    groups.retain(|group| !group.is_empty() && seen.insert(group.clone()));
    Ok(groups)
}

pub fn validate_packages(context: &Context, groups: &[String]) -> Result<(), String> {
    for group in groups {
        for package in directories(&context.root.join(group))? {
            let name = package
                .file_name()
                .and_then(|name| name.to_str())
                .ok_or_else(|| {
                    format!(
                        "package directory name is not valid UTF-8: {}",
                        package.display()
                    )
                })?;
            if name != "overrides" {
                validate_package(name).map_err(|_| {
                    format!("package directory has an unsupported name: {group}/{name}")
                })?;
            }
        }
    }
    Ok(())
}

pub fn render(
    context: &Context,
    groups: &[String],
    metadata: &BTreeMap<String, String>,
) -> Result<(String, String), String> {
    let mut config = String::new();
    let mut document = String::new();
    let mut wrote_group = false;
    for group in groups {
        let mut packages = Vec::new();
        for path in directories(&context.root.join(group))? {
            let name = path
                .file_name()
                .and_then(|name| name.to_str())
                .ok_or_else(|| {
                    format!(
                        "package directory name is not valid UTF-8: {}",
                        path.display()
                    )
                })?;
            if name != "overrides" {
                packages.push(name.to_string());
            }
        }
        if packages.is_empty() {
            continue;
        }
        if wrote_group {
            config.push('\n');
        }
        config.push_str(group);
        config.push_str(" {\n");
        document.push_str("\n## `");
        document.push_str(group);
        document.push_str("`\n\n");
        let width = packages
            .iter()
            .filter(|package| {
                metadata
                    .get(&format!("{group}/{package}"))
                    .is_some_and(|description| !description.is_empty())
            })
            .map(String::len)
            .max()
            .unwrap_or(0)
            + 2;
        for package in packages {
            let description = metadata
                .get(&format!("{group}/{package}"))
                .map(String::as_str)
                .unwrap_or_default();
            config.push_str("  ");
            config.push_str(&package);
            if !description.is_empty() {
                config.push_str(&" ".repeat(width.saturating_sub(package.len())));
                config.push_str("= ");
                config.push_str(description);
            }
            config.push('\n');
            document.push_str("- `");
            document.push_str(&package);
            document.push('`');
            if !description.is_empty() {
                document.push_str(" — ");
                document.push_str(description);
            }
            document.push('\n');
        }
        config.push_str("}\n");
        wrote_group = true;
    }
    if config.ends_with('\n') {
        config.pop();
    }
    Ok((config, document))
}

fn collect_named_files(
    directory: &Path,
    name: &str,
    found: &mut Vec<PathBuf>,
) -> Result<(), String> {
    let entries = match fs::read_dir(directory) {
        Ok(entries) => entries,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(()),
        Err(error) => return Err(format!("read {}: {error}", directory.display())),
    };
    for entry in entries {
        let entry = entry.map_err(|error| format!("read {}: {error}", directory.display()))?;
        let path = entry.path();
        let metadata = fs::symlink_metadata(&path)
            .map_err(|error| format!("read {}: {error}", path.display()))?;
        if metadata.is_dir() {
            collect_named_files(&path, name, found)?;
        } else if path.file_name().and_then(|value| value.to_str()) == Some(name) {
            found.push(path);
        }
    }
    Ok(())
}

pub fn validate_group(group: &str) -> Result<(), String> {
    crate::config::validate_relative(group)?;
    if group.is_empty() {
        return Err("empty group".to_string());
    }
    if group
        .chars()
        .all(|character| character.is_ascii_alphanumeric() || "._/-".contains(character))
    {
        Ok(())
    } else {
        Err(format!("invalid group: {group}"))
    }
}

pub fn validate_package(package: &str) -> Result<(), String> {
    crate::config::validate_relative(package)?;
    if package.is_empty() {
        return Err("empty package".to_string());
    }
    if package
        .chars()
        .all(|character| character.is_ascii_alphanumeric() || "._+@-".contains(character))
    {
        Ok(())
    } else {
        Err(format!("invalid package: {package}"))
    }
}
